application/frontend/controllers/MethodController.php

```diff
@@ -0,0 +1,209 @@
+ * @package frontend\controllers
+ */
+class MethodController extends BaseRestController
+{
+    /**
+     * Get list of Recovery Methods for given user
+     * @param string $employeeId
+     * @return Method[]
+     * @throws BadRequestHttpException
+     */
+    public function actionList(string $employeeId): array
+    {
+        $user = User::findOne(['employee_id' => $employeeId]);
+        if ($user === null) {
+            throw new BadRequestHttpException(
+                'User not found for employeeId ' . var_export($employeeId, true),
+                1540491109
+            );
+        }
+
+        return Method::findAll(['user_id' => $user->id, 'verified' => 1]);
+    }
+
+    /**
+     * Retrieve the Method record referenced by the uid and employee_id
+     * @param string $uid
+     * @return Method
+     * @throws NotFoundHttpException
+     */
+    protected function getRequestedMethod($uid)
+    {
+        $employeeId = \Yii::$app->request->getBodyParam('employee_id');
+
+        $user = User::findOne(['employee_id' => $employeeId]);
+
+        $method = Method::findOne(['uid' => $uid, 'user_id' => ($user->id ?? null)]);
+        if ($method === null) {
+            throw new NotFoundHttpException(
+                'method ' . var_export($uid, true)
+                . ' for employee ' . var_export($employeeId, true)
+                . ' not found',
+                1540665086
+            );
+        }
+
+        return $method;
+    }
+
+    /**
+     * View single Recovery Method
+     * @param string $uid
+     * @return Method
+     * @throws NotFoundHttpException
+     */
+    public function actionView($uid)
+    {
+        return $this->getRequestedMethod($uid);
+    }
+
+    /**
+     * Create new unverified method. Also sends verification message.
+     * @return Method
+     * @throws BadRequestHttpException
+     * @throws ConflictHttpException
+     * @throws \Exception
+     */
+    public function actionCreate()
+    {
+        // ensure we don't use expired methods
+        Method::deleteExpiredUnverifiedMethods();
+
+        $value = \Yii::$app->request->post('value');
+        if ($value === null) {
+            throw new BadRequestHttpException(
+                'value is required',
+                1540665086
+            );
+        }
+
+        $employeeId = \Yii::$app->request->getBodyParam('employee_id');
+        if ($employeeId === null) {
+            throw new BadRequestHttpException(
+                'employee_id is required',
+                1540990164
+            );
+        }
+
+        $userId = User::findOne(['employee_id' => $employeeId])->id ?? null;
+        $method = Method::findOne(['value' => $value, 'user_id' => $userId]);
+
+        if ($method === null) {
+            $method = new Method;
+            $method->user_id = $userId;
+            $method->value = mb_strtolower($value);
+        }
+
+        $method->sendVerification();
+
+        if ( ! $method->save()) {
```

Comment:
We may want to make this `$method->save()` call _before_ we call `$method->sendVerification()`, since calling that `sendVerification()` method itself does a save call (which is not obvious when reading this code).

It also just seems like we would create and save the method before sending the user an email for them to verify the method (which, if the save failed, won't exist).